Added a 'link concept card' RTE button

aks681 opened the pull request:
## Explanation
This PR fixes #8018, by adding a new RTE button that allows a creator to link a concept card to the RTE content. Currently, it is active only in topic and skill editors.

## Checklist
- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [ ] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [ ] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python -m scripts.pre_commit_linter` and `python -m scripts.run_frontend_tests`.
- [ ] The PR is made from a branch that's **not** called "develop".
- [ ] The PR has an appropriate "PROJECT: ..." label (Please add this label for the first-pass review of the PR).
- [ ] The PR has an appropriate "CHANGELOG: ..." label (If you are unsure of which label to add, ask the reviewers for guidance).
- [ ] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [ ] The PR addresses the points mentioned in the codeowner checks for the files/folders changed. (See the [codeowner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).)
- [ ] The PR is **assigned** to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer and don't tick this checkbox.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue. Do not only request the review but also add the reviewer as an assignee.

aks681 committed:
Added concept card rte button

aks681 commented:
There is still one issue where the image uploaded in the skill editor is not visible in the skill selector modal, due to context issues. Will fix that soon.
Other than that, it is ready for review.

aks681 committed:
made review changes

aks681 committed:
lint

seanlip commented on core/controllers/skill_editor.py:
```diff
@@ -194,10 +195,16 @@ class SkillDataHandler(base.BaseHandler):
 
     @acl_decorators.open_access
-    def get(self, comma_separated_skill_ids):
+    def get(self, action, comma_separated_skill_ids):
```
Let's not use this "action" pattern. Make a separate handler.

aks681 replied:
Done

seanlip commented on core/templates/dev/head/components/ck-editor-helpers/ck-editor-4-rte.directive.ts:
```diff
@@ -36,16 +36,21 @@ angular.module('oppia').directive('ckEditor4Rte', [
         var _RICH_TEXT_COMPONENTS = RteHelperService.getRichTextComponents();
         var names = [];
         var icons = [];
+        var isLessonRelatedContext = (
```
Variable names should not start with "is". Perhaps contextIsLessonRelated?

aks681 replied:
Done

seanlip reviewed (changes requested):
Thanks @aks681! I don't have many comments. Please fix the images though before submitting :)

seanlip commented:
Also, lint and backend tests are failing.

aks681 committed:
fixed test

seanlip commented on assets/rich_text_components_definitions.ts:
```diff
@@ -112,20 +112,20 @@ export = {
   },
-  "ConceptCard": {
-    "backend_id": "concept-card",
+  "Conceptcard": {
```
I think I understand why you did this, but it does look a little bit weird. How about using "Concept" throughout?

aks681 replied:
Done

aks681 committed:
made review changes and fixed uploaded image issues

aks681 committed:
minor changes

aks681 commented:
@seanlip Addressed the review comments and fixed the image upload issue. 

Updated screenshots:

RTE with the concept card links:

![image](https://user-images.githubusercontent.com/22347970/70176295-40082000-16fe-11ea-896f-15d2399cacd6.png)

The skill selector modal

![image](https://user-images.githubusercontent.com/22347970/70176352-557d4a00-16fe-11ea-91e4-ed59790b030e.png)

The modal that pops up when the link is clicked

![image](https://user-images.githubusercontent.com/22347970/70176395-67f78380-16fe-11ea-90ad-67722a5f80e1.png)

aks681 committed:
Made review changes

aks681 committed:
lint

aks681 committed:
fixed test

aks681 committed:
fixed lint

aks681 committed:
Merge remote-tracking branch 'upstream/develop' into concept-card-rte

aks681 committed:
made review changes

seanlip commented on core/templates/dev/head/services/context.service.ts:
```diff
@@ -104,7 +105,24 @@ angular.module('oppia').factory('ContextService', [
+      // This function is used in cases where the URL does not specify the
+      // correct context for some case. eg: Viewing a skill's concept card on
+      // any page via the RTE.
+      setCustomEntity: function(entityType, entityId) {
```
Should it be setCustomEntityContext?

Also please make an EntityContext domain object rather than using a raw dict. Then you can use entityContext.getId(), entityContext.getType().

aks681 replied:
Done

seanlip reviewed (changes requested):
Thanks! Looks good, just one more comment.

seanlip commented:
Also @vojtechjelinek @vinitamurthi @nithusha21 @sophiewu6 PTAL as codeowners -- thanks!

vojtechjelinek commented on extensions/rich_text_components/richTextComponentsRequires.ts:
```diff
@@ -25,6 +25,9 @@ require(
   'oppia-noninteractive-link.directive.ts');
+require(
+  'rich_text_components/Concept/directives/' +
+  'oppia-noninteractive-concept.directive.ts');
```
Please move this after the Collapsible require to keep the alphabetical ordering.

aks681 replied:
Done

vojtechjelinek commented on extensions/rich_text_components/components_test.py:
```diff
@@ -114,6 +114,21 @@ def test_link_validation(self):
+
+        self.check_validation(
+            components.Concept, valid_items, invalid_items)
+
```
Nit: Move this after `test_collapsible_validation` to keep the alphabetical ordering.

aks681 replied:
Done

vojtechjelinek commented on extensions/rich_text_components/components.py:
```diff
@@ -121,6 +122,10 @@ class Link(BaseRteComponent):
+class Concept(BaseRteComponent):
+    """Class for Concept component."""
+
+
```
Nit: Move this after Collapsible class to keep the alphabetical ordering.

aks681 replied:
Done

vojtechjelinek commented on extensions/rich_text_components/Image/directives/oppia-noninteractive-image.directive.ts:
```diff
@@ -91,7 +91,6 @@ angular.module('oppia').directive('oppiaNoninteractiveImage', [
               ctrl.filepath);
           } catch (e) {
             var additionalInfo = (
-              '\nExploration ID: ' + ContextService.getExplorationId() +
```
Why removing this?

aks681 replied:
Since this function (getExplorationId) throws an error if not called in an exploration editor (as there is no explorationId there).
Also, in cases it is in an exploration editor/viewer, `entityId` would be `explorationId`  and `entityType` would be `exploration`. So, there is no loss of information in removing this.

vojtechjelinek commented on assets/rich_text_components_definitions.ts:
```diff
@@ -109,6 +112,27 @@ export = {
+      },
+      "default_value": ""
+    }]
+  },
```
Please move this after the Collapsible to keep the alphabetical ordering.

aks681 replied:
Done

vojtechjelinek reviewed (approved):
LGTM from code owner's perspective. Thanks! Just a few ordering related nits.

seanlip reviewed (approved):
Thanks! LGTM.

nithusha21 reviewed (approved):
LGTM as a codeowner (For feconf).

seanlip commented:
Hi @bansalnitish, PTAL as codeowner. Thanks!

seanlip commented:
@bansalnitish is out. Per discussion with @aks681, merging this.
